Merge up a fork?

wesleytodd opened the pull request:
Typically I hate doing this, but I have been using `migrate` on a few projects over the past few years, and have not really seen much work being done in here.  I would be happy to take over some portion of the maintenance of this project, but the things I need/want I am working on in this fork.

I am not sure, but it looks like @LinusU is the main maintainer at the moment.  So @LinusU do you wants some help?  Or do you like the look of the work I am doing in the fork?  I would much rather upstream my work, but if that is not possible maybe we can come up with a way to share work by breaking some core things out into shared modules.

In a technical note, this fork merges almost all the ideas from the open PR's and issues here.  It also adds cli integration tests, standard linting, more robust cli tooling via `commander`, a few new commands and flags, and a general refactors that are mainly personal preference.  Anything in my fork is up for discussion if you think we could actually upstream the work.  So feel free to comment away.

NOTE: this is a WIP branch, so there is much more I plan on doing in here for cleanup, testing, and features.

LinusU commented:
Wow! This looks awesome, I'm on vacation at the moment so won't be able to take a proper look right now, but I hope to do it soon!

wesleytodd commented:
Ok, my fork seems to be diverging even more.  Today I added tracking of migrations by title and their status with a timestamp.  When working on multiple branches that might get merged out of order, the old method of position tracking is broken.  This new addition keeps a timestamp for when the migration was last run up, and consecutive ups will run any migration that hasn't yet run, warning if it ends up running out of order.  This warning means that you might want to run a `clean` up in development, but if running the tool against a production database it means you should probably manually verify the results.

wesleytodd commented:
I made another fairly large change today.  I moved the default `.migrate` file out of the migrations directory, and converted the tool to use `minimatch` glob patterns to filter.  

I made this change to enable you to have your migrations be directories where your `index.js` is required in, and your sql or other related files can live side by side in the directory.

wesleytodd commented:
More additions:

Added an `init` method to the migration state store.  This is run in `migrate init` and `migrate up --init`.   It is used to setup anything the store needs, like a database or directory, to save the migrations state to.

wesleytodd commented:
Hey, I am sorry if I am being a pest, but if I could get some feedback on this work before I go way too far, it would really help me determine if I should be considering the status of this fork merge PR, or if I can drastically re-work some of the features of this. 

So far I have tried to not break all the old ways, but IMO, the cleanest api would require a serious change to the cli flags and options.  I don't want to make my master branch diverge too far if there is a chance of this being merged in.

Anyway, @tj & @LinusU I would love some feedback :)

tj commented:
Oops sorry missed this! I added you as a contrib to the repo/npm, feel free to do what you think it needs! I don't use this package anymore

wesleytodd commented:
No worries @tj. Thanks!

wesleytodd commented:
Ok, I opened the first round of PR's.  Once #73 #74 #72 are merged it will make it much easier to see what this fork does moving forward.  Then we can decide how go about pulling in the rest, because it might make more sense as one big single commit since I really moved a bunch of things around.

motet-a commented:
Any news? This one looks pretty good!

motet-a commented on bin/migrate-list:
```diff
@@ -0,0 +1,57 @@
+
+var program = require('commander')
+var path = require('path')
+var moment = require('moment')
```
What about using a light library like `dateformat` rather than `moment` (which is pretty huge)?

wesleytodd commented:
Sorry I have been busy with vacation and then catching back up after.  I have started the process of opening up prs for this stuff, which are all linked here.  If you want to start using it as is, with the risk of some things changing still before we publish this official one, my fork is published as `@wesleytodd/migrate`.  Also, the branch that this PR is off of is up to date and functional AFAIK.

I hope to be able to get back to this asap, so we can get it all included and published.

RE: moment.  I am fine with that change, being on the server side I did not really put thought to the "light"-ness of the libraries.  But I think that is a good idea.  Will update in one of the subsequent PRs.

motet-a commented:
No worries :wink: 

Since I wrote a lot of client-side stuff these days, I am probably a bit too obsessed with the weight of the libraries.

(By the way, since the situation is currently a bit confused between the `master` branch and this interesting PR, I ended up writing [my own library](https://github.com/motet-a/migrate). I hope this quick hack will not last long, but you can take your time :relieved:)
